### PR TITLE
Realtime profiler (HDRP & URP)

### PR DESCRIPTION
**Pull request by arttu-peltonen**
### Purpose of this PR
This PR introduces the [Realtime Profiler](https://docs.google.com/document/d/1bMTYUb61AOLOCQzdtovhUkoHxQDmWLS8MLIBv8GC2TY/edit?usp=drive_web&ouid=106139378353907605764), which allows users to see cpu/gpu frame timings and bottlenecks detected out of these timings.

profiler2

---

#### Full description of the features with screenshots in [this document](https://docs.google.com/document/d/1v-U0Z5YWmPxnDphrUrEsf59yWsCNWp72q0QD55YwVM0/edit?usp=sharing).

This PR depends on [New FrameTiming Backend PR](https://ono.unity3d.com/unity/unity/pull-request/132585/_/profiler/feature/universal-frame-timing-v3) by @bandures, ~which hasn't landed in trunk yet. Therefore all testing & Yamato runs are done with the latest commit from `trunk/profiler/feature/universal-frame-timing-v3`.~ EDIT: Trunk PR has landed, works at latest trunk now.

---
### Testing status (status mid-September)

Matrix describing what I've tested:

Platform | GFX API | Build type | Tested? | Status 
------------ | ------------- | ------------- | ------------- | ------------- 
Windows | DX11 | Editor | :heavy_check_mark: | :+1:
Windows | DX11 | Player | :heavy_check_mark: | :+1:
Windows | DX12 | Editor | :heavy_check_mark: | :+1:
Windows | DX12 | Player | :heavy_check_mark: | :+1:
Windows | Vulkan | Editor | :heavy_check_mark: | :+1:
Windows | Vulkan | Player | :heavy_check_mark: | :+1:
Mac | Metal | Editor | :heavy_check_mark: | Render thread & GPU timing missing
Mac | Metal | Player | :heavy_check_mark:  | GPU timing missing
Mac | GL | Editor | :heavy_check_mark:  | GPU timing missing
Mac | GL | Player | :heavy_check_mark:  | GPU timing missing, intermittent invalid main thread times
Android | Vulkan | Player | :heavy_check_mark: | :+1:
Android | GLES3 | Player | :heavy_check_mark: | GPU timing missing (Google Pixel 2)
iOS |  | Player | :x:
PS4 |  | Player | :heavy_check_mark: | Render thread timing missing
PS5 |  | Player | :x:
X1 |  | Player | :x:
XSX |  | Player | :x:
Switch |  | Player | :x:
VR/XR |  | Editor | :x:
VR/XR |  | Player | :x:

Tested functionality on HDRP and URP, in editor, editor playmode and player builds.

### CODE REVIEW

**Review by TheoWong-pixel — APPROVED** [review]

File: com.unity.render-pipelines.core/Runtime/Debugging/DebugUI.Containers.cs [hunk 1]
```diff
@@ -53,14 +55,19 @@ public Container(string displayName, ObservableList<Widget> children)
...
+                    OnItemAdded(this.children, new ListChangedEventArgs<Widget>(i, this.children[i]));
             }
 
             internal override void GenerateQueryPath()
             {
                 base.GenerateQueryPath();
 
-                foreach (var child in children)
-                    child.GenerateQueryPath();
+                int numChildren = children.Count;
```
Comment (JulienIgnace-Unity) [inline com.unity.render-pipelines.core/Runtime/Debugging/DebugUI.Containers.cs:68]: I see this pattern being changed in several places here. I'm curious, why is this needed?

**Review by JulienIgnace-Unity — APPROVED** [review]

File: com.unity.render-pipelines.core/Runtime/Debugging/DebugFrameTiming.cs [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        const string k_MsFormatString = "{0:F2}ms";
+        const float k_RefreshRate = 1f / 5f;
+
+        internal FrameTimeSampleHistory m_FrameHistory = new();
+        internal BottleneckHistory m_BottleneckHistory = new();
+
+        /// <summary>
+        /// Size of the Bottleneck History Window in number of samples.
+        /// </summary>
+        public int BottleneckHistorySize { get; set; } = 60;
```
Comment (alex-vazquez) [inline com.unity.render-pipelines.core/Runtime/Debugging/DebugFrameTiming.cs:22]:
```suggestion
        public int bottleneckHistorySize { get; set; } = 60;
```

https://ono.unity3d.com/unity-extra/unity-meta/raw/@/ReferenceSource/CSharp/Assets/CSharpReference.cs

```
// | In the UnityEngine and UnityEditor namespaces (old conventions), public properties are camelCase with no prefix  [FDG Exception]
// | In the Unity namespace, public properties are PascalCase with no prefix
```

File: com.unity.render-pipelines.core/Runtime/Debugging/DebugFrameTiming.cs [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+        /// <summary>
+        /// Size of the Sample History Window in number of samples.
+        /// </summary>
+        public int SampleHistorySize { get; set; } = 30;
```
Comment (alex-vazquez) [inline com.unity.render-pipelines.core/Runtime/Debugging/DebugFrameTiming.cs:27]: Same here, properties start with lowercase
```suggestion
        public int sampleHistorySize { get; set; } = 30;
```

File: com.unity.render-pipelines.core/Runtime/Debugging/DebugManager.cs [hunk 4]
```diff
@@ -241,20 +249,34 @@ void CheckPersistentCanvas()
...
+                    // When container is toggled, we make sure that if there are ValueTuples, they all get the same element index.
+                    int pinnedIndex = container.children.Max(w => (w as DebugUI.ValueTuple)?.pinnedElementIndex ?? -1);
+                    foreach (var child in container.children)
+                    {
+                        if (child is DebugUI.Value || child is DebugUI.ValueTuple)
+                            TogglePersistent(child, pinnedIndex);
+                    }
+                    break;
+                default:
+                    Debug.Log("Only readonly items can be made persistent.");
```
Comment (alex-vazquez) [inline com.unity.render-pipelines.core/Runtime/Debugging/DebugManager.cs:277]: I wonder if this should be better a `warning`

File: com.unity.render-pipelines.core/Runtime/Debugging/DebugUI.cs [hunk 5]
```diff
@@ -235,11 +244,85 @@ public Value()
...
+            /// <summary>
+            /// Returns the formatted value string for display purposes.
+            /// </summary>
+            /// <param name="value">Value to be formatted.</param>
+            /// <returns>The formatted value string.</returns>
+            public virtual string FormatString(object value)
+            {
+                if (formatString != null)
+                    return String.Format(formatString, value);
+                return $"{value}";
```
Comment (alex-vazquez) [inline com.unity.render-pipelines.core/Runtime/Debugging/DebugUI.cs:262]:
```suggestion
                return string.IsNullOrEmpty(formatString) ? $"{value}" : string.Format(formatString, value);
```

**Review by alex-vazquez — APPROVED** [review]
Nice job :)

File: com.unity.render-pipelines.core/Runtime/Debugging/DebugUI.Containers.cs (hunk 1 above)
Comment (arttu-peltonen) [reply com.unity.render-pipelines.core/Runtime/Debugging/DebugUI.Containers.cs:68]: This is an optimization. `UnityEngine.Rendering.ObservableList` implementation is such that using foreach to iterate causes allocations. I wanted to get rid of them, so I removed the usage of foreach. A better fix would be to improve the implementation of ObservableList in this respect, but I didn't want to get into it in this PR.

**Comment by arttu-peltonen** [comment]
@Unity-Technologies/gfx-qa-urp let me know if you're able to handle testing for all URP, HDRP and XR. If not, let's have a chat and ask for help from other QA teams.

File: com.unity.render-pipelines.core/Runtime/Debugging/DebugFrameTiming.cs (hunk 2 above)
Comment (arttu-peltonen) [reply com.unity.render-pipelines.core/Runtime/Debugging/DebugFrameTiming.cs]: Thanks, fixed.

File: com.unity.render-pipelines.core/Runtime/Debugging/DebugFrameTiming.cs (hunk 3 above)
Comment (arttu-peltonen) [reply com.unity.render-pipelines.core/Runtime/Debugging/DebugFrameTiming.cs]: Thanks, fixed.

File: com.unity.render-pipelines.core/Runtime/Debugging/DebugManager.cs (hunk 4 above)
Comment (arttu-peltonen) [reply com.unity.render-pipelines.core/Runtime/Debugging/DebugManager.cs:308]: I'll leave it as Log as it's reaction to user input, it's just info that it's not supported.

File: com.unity.render-pipelines.core/Runtime/Debugging/DebugUI.cs (hunk 5 above)
Comment (arttu-peltonen) [reply com.unity.render-pipelines.core/Runtime/Debugging/DebugUI.cs]: Thanks, fixed.

File: TestProjects/HDRP_Tests/ProjectSettings/ProjectSettings.asset [hunk 6]
```diff
@@ -153,7 +153,7 @@ PlayerSettings:
     enable360StereoCapture: 0
   isWsaHolographicRemotingEnabled: 0
   protectGraphicsMemory: 0
-  enableFrameTimingStats: 0
+  enableFrameTimingStats: 1
```
Comment (jenniferd-unity) [inline TestProjects/HDRP_Tests/ProjectSettings/ProjectSettings.asset:156]: could it slow down our tests? if so, is there a risk of timeout?

File: com.unity.render-pipelines.core/Editor/Debugging/DebugUIDrawer.Builtins.cs [hunk 7]
```diff
@@ -21,7 +21,71 @@ public override bool OnGUI(DebugUI.Widget widget, DebugState state)
...
+    /// </summary>
+    [DebugUIDrawer(typeof(DebugUI.ValueTuple))]
+    public sealed class DebugUIDrawerValueTuple : DebugUIDrawer
+    {
+        /// <summary>
+        /// OnGUI implementation for ValueTuple DebugUIDrawer.
+        /// </summary>
+        /// <param name="widget">DebugUI Widget.</param>
+        /// <param name="state">Debug State associated with the Debug Item.</param>
+        /// <returns>The state of the widget.</returns>
```
Comment (jenniferd-unity) [inline com.unity.render-pipelines.core/Editor/Debugging/DebugUIDrawer.Builtins.cs:41]: in the doc - regarding return value, what does true/false mean exactly? is it the disabled state or something else?

File: com.unity.render-pipelines.core/Editor/Debugging/DebugUIDrawer.Builtins.cs [hunk 8]
```diff
@@ -21,7 +21,71 @@ public override bool OnGUI(DebugUI.Widget widget, DebugState state)
...
+        public override bool OnGUI(DebugUI.Widget widget, DebugState state)
+        {
+            var w = Cast<DebugUI.ValueTuple>(widget);
+
+            var labelRect = PrepareControlRect();
+            EditorGUI.PrefixLabel(labelRect, EditorGUIUtility.TrTextContent(w.displayName));
+
+            // Following layout should match DebugUIDrawerFoldout to make column labels align
+            Rect drawRect = GUILayoutUtility.GetLastRect();
+            const int oneColumnWidth = 70;
```
Comment (jenniferd-unity) [inline com.unity.render-pipelines.core/Editor/Debugging/DebugUIDrawer.Builtins.cs:51]: should this number be in a Styles structure?

File: com.unity.render-pipelines.core/Editor/Debugging/DebugUIDrawer.Builtins.cs [hunk 9]
```diff
@@ -21,7 +21,71 @@ public override bool OnGUI(DebugUI.Widget widget, DebugState state)
...
+            {
+                var columnRect = drawRect;
+                columnRect.x += EditorGUIUtility.labelWidth + i * oneColumnWidth;
+                columnRect.width = oneColumnWidth;
+                var value = w.values[i].GetValue();
+                EditorGUI.LabelField(columnRect, w.values[i].FormatString(value));
+            }
+            EditorGUI.indentLevel = indent;
+
+            return true;
```
Comment (jenniferd-unity) [inline com.unity.render-pipelines.core/Editor/Debugging/DebugUIDrawer.Builtins.cs:64]: wait we always return true here - am I missing a path where we can return false?

File: com.unity.render-pipelines.core/Editor/Debugging/DebugUIDrawer.Builtins.cs [hunk 10]
```diff
@@ -21,7 +21,71 @@ public override bool OnGUI(DebugUI.Widget widget, DebugState state)
...
+    /// </summary>
+    [DebugUIDrawer(typeof(DebugUI.ProgressBarValue))]
+    public sealed class DebugUIDrawerProgressBarValue : DebugUIDrawer
+    {
+        /// <summary>
+        /// OnGUI implementation for Value DebugUIDrawer.
+        /// </summary>
+        /// <param name="widget">DebugUI Widget.</param>
+        /// <param name="state">Debug State associated with the Debug Item.</param>
+        /// <returns>The state of the widget.</returns>
```
Comment (jenniferd-unity) [inline com.unity.render-pipelines.core/Editor/Debugging/DebugUIDrawer.Builtins.cs:79]: same remarks as for DebugUIDrawerValueTuple

File: com.unity.render-pipelines.core/Runtime/Debugging/DebugFrameTiming.cs [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+
+//#define RTPROFILER_DEBUG
+
+namespace UnityEngine.Rendering
+{
+    public class DebugFrameTiming
+    {
+        const string k_FpsFormatString = "{0:F1}";
+        const string k_MsFormatString = "{0:F2}ms";
+        const float k_RefreshRate = 1f / 5f;
```
Comment (jenniferd-unity) [inline com.unity.render-pipelines.core/Runtime/Debugging/DebugFrameTiming.cs:14]: why this value for the refresh rate?

File: com.unity.render-pipelines.core/Runtime/Debugging/DebugManager.cs [hunk 12]
```diff
@@ -264,20 +280,34 @@ void EnsurePersistentCanvas()
...
```
Comment (jenniferd-unity) [inline com.unity.render-pipelines.core/Runtime/Debugging/DebugManager.cs:308]: should we keep the debug log here? i assumed this will show up in console and i am not sure what the users are supposed to do with it when they see it

File: com.unity.render-pipelines.core/Runtime/Debugging/Prefabs/Scripts/DebugUIHandlerEnumHistory.cs [hunk 13]
```diff
@@ -9,17 +9,18 @@ namespace UnityEngine.Rendering.UI
     public class DebugUIHandlerEnumHistory : DebugUIHandlerEnumField
     {
         Text[] historyValues;
-        const float xDecal = 60f;
+        const float k_XOffset = 230f;
```
Comment (jenniferd-unity) [inline com.unity.render-pipelines.core/Runtime/Debugging/Prefabs/Scripts/DebugUIHandlerEnumHistory.cs:12]: should it be in its own DebugPanel Styles class?

File: com.unity.render-pipelines.core/Runtime/Debugging/Prefabs/Scripts/DebugUIHandlerFoldout.cs [hunk 14]
```diff
@@ -15,8 +15,8 @@ public class DebugUIHandlerFoldout : DebugUIHandlerWidget
         DebugUI.Foldout m_Field;
         DebugUIHandlerContainer m_Container;
 
-        const float xDecal = 60f;
-        const float xDecalInit = 230f;
+        const float k_FoldoutXOffset = 215f;
```
Comment (jenniferd-unity) [inline com.unity.render-pipelines.core/Runtime/Debugging/Prefabs/Scripts/DebugUIHandlerFoldout.cs:18]: same remark with Styles class

File: com.unity.render-pipelines.core/Runtime/Debugging/Prefabs/Scripts/DebugUIHandlerValue.cs [hunk 15]
```diff
@@ -11,9 +11,10 @@ public class DebugUIHandlerValue : DebugUIHandlerWidget
         public Text nameLabel;
         /// <summary>Value of the value field.</summary>
         public Text valueLabel;
-        DebugUI.Value m_Field;
 
-        float m_Timer;
+        DebugUI.Value m_Field;
+        protected internal float m_Timer;
+        static readonly Color k_ZeroColor = new Color(0.5f, 0.5f, 0.5f, 1f);
```
Comment (jenniferd-unity) [inline com.unity.render-pipelines.core/Runtime/Debugging/Prefabs/Scripts/DebugUIHandlerValue.cs:17]: i would recommend using Color.gray instead

File: com.unity.render-pipelines.core/Runtime/Debugging/Prefabs/Scripts/DebugUIHandlerValueTuple.cs [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+        public Text nameLabel;
+        /// <summary>Value of the value field.</summary>
+        public Text valueLabel;
+
+        protected internal DebugUI.ValueTuple m_Field;
+        protected internal Text[] valueElements;
+
+        const float k_XOffset = 230f;
+        float m_Timer;
+        static readonly Color k_ZeroColor = new Color(0.5f, 0.5f, 0.5f, 1f);
```
Comment (jenniferd-unity) [inline com.unity.render-pipelines.core/Runtime/Debugging/Prefabs/Scripts/DebugUIHandlerValueTuple.cs:23]: same remark as above, please use Color.gray

File: com.unity.render-pipelines.core/Tests/Runtime/RuntimeProfilerTests.cs [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+            for (int i = 0; i < k_NumWarmupFrames; i++)
+                yield return new WaitForEndOfFrame();
+
+            m_DebugFrameTiming.Reset();
+        }
+    }
+
+    // FIXME: Tests are disabled in player builds for now, since there's no API that tells whether frame timing is
+    //        enabled or not. Re-enable if that changes.
+#if true//UNITY_EDITOR
```
Comment (jenniferd-unity) [inline com.unity.render-pipelines.core/Tests/Runtime/RuntimeProfilerTests.cs:59]: do you have a fogbugz or jira for this? so we dont forget :)

**Review by jenniferd-unity — APPROVED** [review]
Approved! great job @arttu-peltonen !
I have added a few comments I am sure you will know how to tackle

File: TestProjects/HDRP_Tests/ProjectSettings/ProjectSettings.asset (hunk 6 above)
Comment (arttu-peltonen) [reply TestProjects/HDRP_Tests/ProjectSettings/ProjectSettings.asset:156]: It shouldn't. When this flag is enabled, it will affect master builds too - it enables a some frame-level timing stats on the backend. These are the timings that could be used to implement dynamic resolution scaling logic. It's not zero overhead but should be very small. @bandures might have some numbers on this.

File: com.unity.render-pipelines.core/Editor/Debugging/DebugUIDrawer.Builtins.cs (hunk 7 above)
Comment (arttu-peltonen) [reply com.unity.render-pipelines.core/Editor/Debugging/DebugUIDrawer.Builtins.cs:41]: It's "is container open" state, so for non-container widgets it doesn't mean anything. This is how all other widgets are documented so I'm following the form here.

File: com.unity.render-pipelines.core/Editor/Debugging/DebugUIDrawer.Builtins.cs (hunk 9 above)
Comment (arttu-peltonen) [reply com.unity.render-pipelines.core/Editor/Debugging/DebugUIDrawer.Builtins.cs:64]: Nope, the return value doesn't have an effect here because this is not a container widget. For container widgets, the returned value would control whether or not the child widgets should be drawn.

File: com.unity.render-pipelines.core/Runtime/Debugging/DebugFrameTiming.cs (hunk 11 above)
Comment (arttu-peltonen) [reply com.unity.render-pipelines.core/Runtime/Debugging/DebugFrameTiming.cs:14]: It's the same value that [HDRP had](https://github.com/Unity-Technologies/Graphics/pull/5419/files#diff-9c446cfe3ccbc91b147ecf9663bcff3908953ec24bce72c3c7b3605ccc531b14L971) previously, so I didn't change it. I guess the refresh rate is a compromise between performance (it costs something to update the value), readability (too quickly changing values can be hard to read) and correctness (how up to date the value is). I think this value is a reasonable compromise.

File: com.unity.render-pipelines.core/Runtime/Debugging/DebugManager.cs (hunk 12 above)
Comment (arttu-peltonen) [reply com.unity.render-pipelines.core/Runtime/Debugging/DebugManager.cs:308]: Yeah it will show up in the console. I think it's good to keep it because the way to make values persistent (pin debug value to screen) is just RShift, and there is no visual indication whether this action is supported for the selected debug value. So for readonly values pressing RShift will pin the value, and for other values where this is not supported, pressing RShift will print this message to console. This way at least there's at least a chance that user will see the message and understand why nothing happened on screen.

File: com.unity.render-pipelines.core/Runtime/Debugging/Prefabs/Scripts/DebugUIHandlerEnumHistory.cs (hunk 13 above)
Comment (arttu-peltonen) [reply com.unity.render-pipelines.core/Runtime/Debugging/Prefabs/Scripts/DebugUIHandlerEnumHistory.cs:12]: Perhaps. I don't feel like refactoring this code too much (especially the runtime part) because I think we should replace this UGUI code with UITK.

File: com.unity.render-pipelines.core/Tests/Runtime/RuntimeProfilerTests.cs (hunk 17 above)
Comment (arttu-peltonen) [reply com.unity.render-pipelines.core/Tests/Runtime/RuntimeProfilerTests.cs:59]: I'll make sure to add a ticket. Also thanks for highlighting this line, `#if EDITOR` should be enabled (because as the comment explains, in Player we don't yet know if the data will be valid or not).

File: com.unity.render-pipelines.core/Editor/Debugging/DebugUIDrawer.Builtins.cs (hunk 8 above)
Comment (arttu-peltonen) [reply com.unity.render-pipelines.core/Editor/Debugging/DebugUIDrawer.Builtins.cs:51]: Done!

**Comment by arttu-peltonen** [comment]
Update: Added Performance Analysis section to the [PR notes doc](https://docs.google.com/document/d/1v-U0Z5YWmPxnDphrUrEsf59yWsCNWp72q0QD55YwVM0/edit#heading=h.8lr3rnxm2j9j).

File: com.unity.render-pipelines.core/Runtime/Debugging/DebugUI.cs [hunk 18]
```diff
@@ -235,11 +244,83 @@ public Value()
...
             }
+
+            /// <summary>
+            /// Returns the formatted value string for display purposes.
+            /// </summary>
+            /// <param name="value">Value to be formatted.</param>
+            /// <returns>The formatted value string.</returns>
+            public virtual string FormatString(object value)
+            {
+                return string.IsNullOrEmpty(formatString) ? $"{value}" : string.Format(formatString, value);
```
Comment (Skiri-ki) [inline com.unity.render-pipelines.core/Runtime/Debugging/DebugUI.cs:260]: I guess this would be a good place to switch to [MutableString](https://docs.google.com/document/d/1MUWvnECRWz71FymdOtSoJ_Z8cj7-oLcTuijhg-jVfYs) once available...

**Review by sebastienlagarde — APPROVED** [review]

**Comment by fabien-unity** [comment]
I quickly tried with XR and the report seems accurate.
However, the debug display is only visible in the mirror view (desktop). This is something not related to this PR but we should think about improving the debug render to be compatible with XR inside the headset.

File: com.unity.render-pipelines.core/Runtime/Debugging/DebugUI.cs (hunk 18 above)
Comment (arttu-peltonen) [reply com.unity.render-pipelines.core/Runtime/Debugging/DebugUI.cs:260]: Yes, absolutely. I've spent some time trying to eliminate allocs but this was the one place I couldn't easily fix without removing support for format strings altogether, which I would like to keep. Great to hear that there's work for a non-allocating StringBuilder/String.Format that I could use in the future!

**Comment by arttu-peltonen** [comment]
> I quickly tried with XR and the report seems accurate.
> However, the debug display is only visible in the mirror view (desktop). This is something not related to this PR but we should think about improving the debug render to be compatible with XR inside the headset.

Thanks! Yeah, I agree. I have only tested with MockHMD, not headset, and haven't investigated how XR is supported with uGUI.

**Comment by arttu-peltonen** [comment]
Note: there's a new version of the [trunk PR](https://ono.unity3d.com/unity/unity/pull-request/132585/_/profiler/feature/universal-frame-timing-v3) now, so please use `profiler/feature/universal-frame-timing-v3` branch for testing.

**Review by eh-unity — APPROVED** [review]
Looks good to me.

**Review by aleksandrasdzikia — APPROVED** [review]
Looks good to me. Tested on two different macs (intel and apple silicon). Both behaved as expected. Apple silicon on Rosetta had issues, but not related to this PR. tested both Metal and GL. Different URP projects were used - Universal Rendering Examples, URP Template, Viking Village. Behaviour of the frame stats were accurate to the Display stats. 

Overall, very useful tool. Only one suggestion I have is from UI/UX perspective. It would make sense to not hide "Display Stats" if it's not in Play Mode. I would add some info there like: "Please, press play to display the stats" or something similar. Otherwise, it might look confusing if it's the first time of using it. Other than that, looks good!

**Review by phi-lira — APPROVED** [review]
Approving on yamato state. Terrain failures are known issue on master.

**Review by Verasl — APPROVED** [review]
LGTM
